blobstore-s3/src/main.rs

```diff
@@ -0,0 +1,193 @@
+//! main process for blobstore-s3 provider
+//! This is a thin wrapper around the blobstore-s3 library
+//!
+use blobstore_s3_lib::{
+    wasmcloud_interface_blobstore::{
+        Blobstore, BlobstoreReceiver, ContainerId, ContainerIds, ContainerMetadata,
+        ContainerObject, ContainersInfo, GetObjectRequest, GetObjectResponse, ListObjectsRequest,
+        ListObjectsResponse, MultiResult, ObjectMetadata, PutChunkRequest, PutObjectRequest,
+        PutObjectResponse, RemoveObjectsRequest,
+    },
+    StorageClient, StorageConfig,
+};
+use std::{collections::HashMap, convert::Infallible, sync::Arc};
+use tokio::sync::RwLock;
+use wasmbus_rpc::{core::LinkDefinition, provider::prelude::*};
+
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    // handle lattice control messages and forward rpc to the provider dispatch
+    // returns when provider receives a shutdown control message
+    provider_main(S3BlobstoreProvider::default())?;
+
+    eprintln!("blobstore-s3 provider exiting");
+    Ok(())
+}
+
+/// Nats implementation for wasmcloud:messaging
+#[derive(Default, Clone, Provider)]
+#[services(Blobstore)]
+struct S3BlobstoreProvider {
+    // store nats connection client per actor
+    actors: Arc<RwLock<HashMap<String, StorageClient>>>,
+}
+
+// use default implementations of provider message handlers
+impl ProviderDispatch for S3BlobstoreProvider {}
+
+impl S3BlobstoreProvider {
+    async fn client(&self, ctx: &Context) -> RpcResult<StorageClient> {
+        let actor_id = ctx
+            .actor
+            .as_ref()
+            .ok_or_else(|| RpcError::InvalidParameter("no actor in request".to_string()))?;
+        // get read lock on actor-client hashmap
+        let rd = self.actors.read().await;
+        let client = rd
+            .get(actor_id)
+            .ok_or_else(|| RpcError::InvalidParameter(format!("actor not linked:{}", actor_id)))?;
+        Ok(client.clone())
+    }
+
+    // s3 client: like client() but doesn't clone linkdefinition
+    // (possibly a premature optimization!)
+    async fn s3_client(&self, ctx: &Context) -> RpcResult<StorageClient> {
```

Comment:
I feel like if it's just saving us a clone of the `LinkDefinition` structure it likely isn't worth two different codepaths. I could be convinced though